src/org/etools/j1939_84/controllers/part07/Part07Step18Controller.java

```diff
@@ -54,21 +57,81 @@
+        incrementProgress("Step 6.7.18.1.e - Turn Engine On and keep the ignition key in the on position");
+        ensureKeyOnEngineOn();
+
         // 6.7.18.1.f. Wait for manufacturer’s recommended time for Fault B to be detected as failed.
+        waitForFault("Step 6.7.18.1.f");
+
         // 6.7.18.1.g. Turn engine off.
+        incrementProgress("Step 6.7.18.1.g - Turn Engine Off and keep the ignition key in the off position");
+        ensureKeyOffEngineOff();
+
         // 6.7.18.1.h. Wait engine manufacturer’s recommended interval for permanent fault recording.
+        waitForManufacturerInterval("Step 6.7.18.1.h", "off");
+
         // 6.7.18.1.i. Start Engine.
+        incrementProgress("Step 6.7.18.1.i - Turn Engine on with the ignition key in the on position");
+        ensureKeyOnEngineOn();
+
         // 6.7.18.1.j. If Fault B is a single trip fault proceed with part 8 immediately.
-        // 6.7.18.1.k. Wait for manufacturer’s recommended time for Fault B to be detected as failed.
-        // 6.7.18.1.l. Turn engine off.
-        // 6.7.18.1.m. Wait engine manufacturer’s recommended interval for permanent fault recording.
-        // 6.7.18.1.n. Start Engine.
-        // 6.7.18.1.o. Proceed with part 8 (cycle 8b).
+        int numberOfTripsForFaultBImplant = getDataRepository().getVehicleInformation()
+                                                               .getNumberOfTripsForFaultBImplant();
+        if (numberOfTripsForFaultBImplant != 1) {
+            // We already handle the first fault B, so only iterate over rest
+            for (int i = 2; i <= numberOfTripsForFaultBImplant; i++) {
+                incrementProgress(format("Step 6.7.18.1.j - Running fault B trip #%d of %d total fault trips",
+                                         i,
+                                         numberOfTripsForFaultBImplant));
+                // 6.7.18.1.k. Wait for manufacturer’s recommended time for Fault B to be detected as failed.
+                waitForFault("Step 6.7.18.1.k");
+
+                // 6.7.18.1.l. Turn engine off.
+                incrementProgress("Step 6.7.18.1.l - Turn Engine Off and keep the ignition key in the off position.");
+                ensureKeyOffEngineOff();
+
+                // 6.7.18.1.m. Wait engine manufacturer’s recommended interval for permanent fault recording.
+                waitForManufacturerInterval("Step 6.7.18.1.m", "off");
+
+                // 6.7.18.1.n. Start Engine.
+                // 6.7.18.1.o. Proceed with part 8 (cycle 8b).
+                incrementProgress("Step 6.7.18.1.n & o - With the ignition key on and engine on proceeding to Part 8");
+                ensureKeyOnEngineOn();
+            }
+        } else {
+            // 6.7.18.1.j. If Fault B is a single trip fault proceed with part 8 immediately.
+            incrementProgress("Step 6.7.18.1.j - Fault B is a single trip fault; proceeding with part 8 immediately");
```

Comment:
This message will not be seen the log, nor will the user actually see it.  But don't remove it.  I think we might want to note on the report the number of trips for Fault B.

Reply:
We do if it is more than 1.  Otherwise, it is not recorded.  See the first line of code in the for loop located near line 97 of this class.